ledger/metrics/src/main/scala/com/daml/metrics/grpc/GrpcMetricsServerInterceptor.scala

```diff
@@ -0,0 +1,142 @@
+      MetricsGrpcClientType -> clientType,
+      MetricsGrpcServerType -> serverType,
+    ) { implicit mc =>
+      val timerHandle = metrics.callTimer.startAsync()
+      metrics.callsStarted.mark()
+      val metricsServerCall = new MetricsCall(
+        serverCall,
+        timerHandle,
+        metrics.messagesSent,
+        metrics.messagesSentSize,
+        metrics.callsFinished,
+      )
+      new MonitoringServerCallListener(
+        serverCallHandler.startCall(metricsServerCall, metadata),
+        metrics.messagesReceived,
+        metrics.messagesReceivedSize,
+      )
+    }
+  }
+
+  private final class MonitoringServerCallListener[T](
+      val delegate: Listener[T],
+      messagesReceived: Meter,
+      messagesReceivedSize: Histogram,
+  )(implicit
+      metricsContext: MetricsContext
+  ) extends ForwardingServerCallListener[T] {
+
+    override def onMessage(message: T): Unit = {
+      super.onMessage(message)
+      updateHistogramWithSerializedSize(messagesReceivedSize, message)
+      messagesReceived.mark()
+    }
+  }
+
+  private final class MetricsCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      timer: TimerHandle,
+      messagesSent: Meter,
+      messagesSentSize: Histogram,
+      callsClosed: Meter,
+  )(implicit metricsContext: MetricsContext)
+      extends SimpleForwardingServerCall[ReqT, RespT](call) {
+
+    override def sendMessage(message: RespT): Unit = {
+      super.sendMessage(message)
+      updateHistogramWithSerializedSize(messagesSentSize, message)
+      messagesSent.mark()
+    }
+
+    override def close(status: Status, trailers: Metadata): Unit = {
+      super.close(status, trailers)
+      withExtraMetricLabels(MetricsGrpcResponseCode -> status.getCode.toString) {
+        implicit metricsContext =>
+          callsClosed.mark()
+          timer.stop()
+      }
+    }
+  }
+
+  private def updateHistogramWithSerializedSize[T](
+      histogram: Histogram,
+      message: T,
+  )(implicit metricsContext: MetricsContext): Unit = {
+    message match {
+      case generatedMessage: JavaGeneratedMessage =>
+        histogram.update(generatedMessage.getSerializedSize)
+      case generatedMessage: ScalapbGeneratedMessage =>
+        histogram.update(generatedMessage.serializedSize)
+      case _ =>
+    }
+  }
+}
+object GrpcMetricsServerInterceptor {
+
+  val MetricsGrpcServiceName = "grpc_service_name"
+  val MetricsGrpcMethodName = "grpc_method_name"
+  val MetricsGrpcClientType = "grpc_client_type"
+  val MetricsGrpcServerType = "grpc_server_type"
+  val MetricsRequestTypeUnary = "unary"
+  val MetricsRequestTypeStreaming = "streaming"
+  val MetricsGrpcResponseCode = "grpc_code"
+
+}
+
+trait GrpcServerMetrics {
+  val callTimer: Timer
+  val messagesSent: Meter
+  val messagesSentSize: Histogram
+  val messagesReceived: Meter
+  val messagesReceivedSize: Histogram
+  val callsStarted: Meter
+  val callsFinished: Meter
```

Comment:
I think there is a need for an `errors` metric (it was in the requirements).
It is a copy of `callsFinished`, but only on error status code.
The idea is that it is way easier for the application to know what is an error, than having to filter in or out the right set of status code in Grafana.

Reply:
@lucb-da What application would be in this case? 
I'm fairly against this because it would add a new series of metrics to prometheus that is an alias of the existing metrics.
If the use wants such a metric it can add a prometheus recording rule to create it.
In fact, the go exporter does the same thing of using the error code https://github.com/grpc-ecosystem/go-grpc-prometheus